docs/configuration.rst
Configuration
=============

thumbor's configuration file is just a regular python script that
gets loaded by thumbor.

In order to get a commented configuration file, just run:

::

    thumbor-config > ./thumbor.conf

Override config through environment variable
-----------------------------------------------

It is possible overide **string configs** through environment variables.
This is possible because thumbor uses `derpconf <https://github.com/globocom/derpconf>`__
to abstract loading configuration and derpconf allows this.

Extensibility Section
---------------------

LOADER
~~~~~~

The loader is responsible for retrieving the source image that thumbor
will work with. This configuration defines the module that thumbor will
use for it. **This must be a full namespace module (a.k.a. python has to
be able to *import* it).**

.. code:: python

   LOADER = 'thumbor.loaders.http_loader'

STORAGE
~~~~~~~

The storage is responsible for storing the source image bytes and
related metadata (face-detection, encryption and such) so that we don't
keep loading it every time. **This must be a full namespace module
(a.k.a. python has to be able to *import* it).**

.. code:: python

   STORAGE = 'thumbor.storages.file_storage'

MIXED\_STORAGE\_FILE\_STORAGE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you are using thumbor's mixed storage
(``thumbor.storages.mixed_storage``), this is where you specify the storage
that will be used to store images. **This must be a full namespace
module (a.k.a. python has to be able to *import* it).**

.. code:: python

   MIXED_STORAGE_FILE_STORAGE = 'thumbor.storages.file_storage'

MIXED\_STORAGE\_CRYPTO\_STORAGE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you are using thumbor's mixed storage
(``thumbor.storages.mixed_storage``), this is where you specify the storage
that will be used to store cryptography information. **This must be a
full namespace module (a.k.a. python has to be able to *import* it).**

.. code:: python

   MIXED_STORAGE_CRYPTO_STORAGE = 'thumbor.storages.file_storage'

MIXED\_STORAGE\_DETECTOR\_STORAGE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you are using thumbor's mixed storage
(``thumbor.storages.mixed_storage``), this is where you specify the storage
that will be used to store facial and feature detection results. **This
must be a full namespace module (a.k.a. python has to be able to
*import* it)**.

.. code:: python

   MIXED_STORAGE_DETECTOR_STORAGE = 'thumbor.storages.file_storage'

RESULT\_STORAGE
~~~~~~~~~~~~~~~

The result storage is responsible for storing the resulting image with
the specified parameters (think of it as a cache), so that we don't keep
processing it every time a request comes in. **This must be a full
namespace module (a.k.a. python has to be able to *import* it).**

.. code:: python

   RESULT_STORAGE = 'thumbor.result_storages.file_storage'

ENGINE
~~~~~~

The engine is responsible for transforming the image. **This must be a
full namespace module (a.k.a. python has to be able to *import* it).**

Currently, thumbor ships with only the ``thumbor.engines.pil`` imaging engine. A few years ago we conducted a comparison between the engines and there was no clear winner. Given PIL was the engine we were using at the time, we decided to stick with it. Other open source engines exist and you can find more about them in the plug-in section of the docs.

.. code:: python

   ENGINE = 'thumbor.engines.pil'

URL\_SIGNER
~~~~~~~~~~~

The url signer is responsible for validation and signing of requests to prevent url tampering,
which could lead to denial of service (example: filling the result_storage by specifying a different size).
**This must be a full namespace module (a.k.a. python has to be able to *import* it).**

.. code:: python

   URL_SIGNER = 'libthumbor.url_signers.base64_hmac_sha1'

Filters Section
---------------

In order to specify the filters that thumbor will use, you need a
configuration key called ``FILTERS``. This is a regular python list with the
full names (names that python can import) of the filter modules you want
to use.

i.e.:

.. code:: python

    FILTERS = [
        'thumbor.filters.brightness',
        'thumbor.filters.contrast',
        'thumbor.filters.rgb',
        'thumbor.filters.round_corner',
        'thumbor.filters.quality',
        'thumbor.filters.noise',
        'thumbor.filters.watermark',
    ]

Metadata Section
----------------

META\_CALLBACK\_NAME
~~~~~~~~~~~~~~~~~~~~

If you want thumbor to use JSONP for image metadata instead of using
JSON, just set this variable to the callback name you want.

.. code:: python

   META_CALLBACK_NAME = 'thumbor_callback'  # Or None for no callback

Face and Feature Detection Section
----------------------------------

DETECTORS
~~~~~~~~~

This options specifies the detectors that should run the image to check
for focal points.

i.e.:

.. code:: python

    DETECTORS = [
      'thumbor.detectors.face_detector',
      'thumbor.detectors.feature_detector'
   ]

Cascade Files
~~~~~~~~~~~~~

This option specifies the cascade (XML) file paths to train openCV to
find faces or other objects.

.. code:: python

   ## The cascade file that opencv will use to detect faces.
   FACE_DETECTOR_CASCADE_FILE = 'haarcascade_frontalface_alt.xml'

   ## The cascade file that opencv will use to detect glasses.
   GLASSES_DETECTOR_CASCADE_FILE = 'haarcascade_eye_tree_eyeglasses.xml'

   ## The cascade file that opencv will use to detect profile faces.
   PROFILE_DETECTOR_CASCADE_FILE = 'haarcascade_profileface.xml'


Imaging Section
---------------

ALLOWED\_SOURCES
~~~~~~~~~~~~~~~~

This configuration defines the source of the images that thumbor will
load. This is only used in the HttpLoader (check the LOADER
configuration above).

.. code:: python

   ALLOWED_SOURCES=['http://s.glbimg.com']

Another example with wildcards:

.. code:: python

   ALLOWED_SOURCES=['.+\.globo\.com', '.+\.glbimg\.com']

This is to get any images that are in ``*.globo.com`` or ``*.glbimg.com`` and it
will fail with any other domains.

MAX\_WIDTH and MAX\_HEIGHT
~~~~~~~~~~~~~~~~~~~~~~~~~~

These define the box that the resulting image for thumbor must fit-in.
This means that no image that thumbor generates will have a width larger
than MAX\_WIDTH or height larger than MAX\_HEIGHT. It defaults to 0, which
means there is not limit. If the original image is larger than
MAX\_WIDTH x MAX\_HEIGHT, it is proportionally resized to MAX\_WIDTH x MAX\_HEIGHT.

.. code:: python

    MAX_WIDTH = 1200
    MAX_HEIGHT = 800

MIN\_WIDTH and MIN\_HEIGHT
~~~~~~~~~~~~~~~~~~~~~~~~~~

These define the box that the resulting image for thumbor must fit-in.
This means that no image that thumbor generates will have a width
smaller than MIN\_WIDTH or height smaller than MIN\_HEIGHT. It defaults to 1.
If the original image is smaller than  MIN\_WIDTH x MIN\_HEIGHT, it is
proportionally resized to MIN\_WIDTH x MIN\_HEIGHT.

.. code:: python

    MIN_WIDTH = 1
    MIN_HEIGHT = 1

QUALITY
~~~~~~~

This option defines the quality that JPEG images will be generated with.
It defaults to 80.

.. code:: python

   QUALITY = 90

MAX\_AGE
~~~~~~~~

This option defines the number of seconds that images should remain in
the browser's cache. It relates directly with the Expires and
Cache-Control headers.

.. code:: python

   MAX_AGE = 24 * 60 * 60  # A day of caching

MAX\_AGE\_TEMP\_IMAGE
~~~~~~~~~~~~~~~~~~~~~

When an image has some error in its detection or it has deferred
queueing, it's convenient to set a much lower expiration time for the
image cache. This way the browser will request the proper image faster.

This option defines the number of seconds that images in this scenario
should remain in the browser's cache. It relates directly with the
Expires and Cache-Control headers.

.. code:: python

   MAX_AGE_TEMP_IMAGE = 60  # A minute of caching

RESPECT\_ORIENTATION
~~~~~~~~~~~~~~~~~~~~

If this option is set to True, thumbor will reorient the image according
to it's EXIF Orientation tag (if one can be found). This options
defaults to False.

The operations performed in the image are as follow (considering the
value of the Orientation EXIF tag):

1. Nothing
2. Flips the image horizontally
3. Rotates the image 180 degrees
4. Flips the image vertically
5. Flips the image vertically and rotates 270 degrees
6. Rotates the image 270 degrees
7. Flips the image horizontally and rotates 270 degrees
8. Rotates the image 90 degrees

.. code:: python

   RESPECT_ORIENTATION = False

ALLOW\_ANIMATED\_GIFS
~~~~~~~~~~~~~~~~~~~~~

This option indicates whether animated gifs should be supported.

.. code:: python

   ALLOW_ANIMATED_GIFS = True

USE\_GIFSICLE\_ENGINE
~~~~~~~~~~~~~~~~~~~~~

This option indicates whether
`gifsicle <http://www.lcdf.org/gifsicle/man.html>`__ should be used for
all gif images, instead of the actual imaging engine. This defaults to
False.

**When using gifsicle thumbor will generate proper animated gifs, as
well as static gifs with the smallest possible size.**

.. code:: python

   USE_GIFSICLE_ENGINE = True

WARNING: When using gifsicle engine, filters will be skipped. thumbor
will not do smart cropping as well.

AUTO\_WEBP
~~~~~~~~~~

This option indicates whether thumbor should send WebP images
automatically if the request comes with an "Accept" header that
specifies that the browser supports "image/webp".

.. code:: python

   AUTO_WEBP = True

AUTO\_PNG\_TO\_JPG
~~~~~~~~~~~~~~~~~~

This option indicates whether thumbor should transform PNG images
automatically to JPEG. If the image is a PNG without transparency and
the numpy dependency is installed, thumbor will transform from png to jpeg.
In the most of cases the image size will decrease.

WARNING: Depending on case, this is not a good deal. This transformation
maybe causes distortions or the size of image can increase.
Images with texts, for example, the result image maybe will be distorced.
Dark images, for example, the size of result image maybe will be bigger.
You have to evaluate the majority of your use cases to take a decision about the usage of this conf.

.. code:: python

   AUTO_PNG_TO_JPG = True

Queueing - Redis
----------------

REDIS\_QUEUE\_SERVER\_HOST
~~~~~~~~~~~~~~~~~~~~~~~~~~

Server host for the queued redis detector.

.. code:: python

   REDIS_QUEUE_SERVER_HOST = 'localhost'

REDIS\_QUEUE\_SERVER\_PORT
~~~~~~~~~~~~~~~~~~~~~~~~~~

Server port for the queued redis detector.

.. code:: python

   REDIS_QUEUE_SERVER_PORT = 6379

REDIS\_QUEUE\_SERVER\_DB
~~~~~~~~~~~~~~~~~~~~~~~~

Server database index for the queued redis detector

.. code:: python

   REDIS_QUEUE_SERVER_DB = 0

REDIS\_QUEUE\_SERVER\_PASSWORD
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Server password for the queued redis detector

.. code:: python

   REDIS_QUEUE_SERVER_PASSWORD = None

Queueing - Amazon SQS
---------------------

This queue will be removed in an upcoming release in favor of the open source AWS plug-ins for thumbor.

SQS\_QUEUE\_KEY\_ID
~~~~~~~~~~~~~~~~~~~

Amazon AWS key id.

.. code:: python

   SQS_QUEUE_KEY_ID = None

SQS\_QUEUE\_KEY\_SECRET
~~~~~~~~~~~~~~~~~~~~~~~

Amazon AWS key secret.

.. code:: python

   SQS_QUEUE_KEY_SECRET = None

SQS\_QUEUE\_REGION
~~~~~~~~~~~~~~~~~~

Amazon AWS SQS region.

.. code:: python

   SQS_QUEUE_REGION = 'us-east-1'

Security Section
----------------

SECURITY\_KEY
~~~~~~~~~~~~~

This option specifies the security key that thumbor uses to sign secure
URLs.

.. code:: python

   1234567890123456

ALLOW\_UNSAFE\_URL
~~~~~~~~~~~~~~~~~~


thumbor/config.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# thumbor imaging service
# https://github.com/thumbor/thumbor/wiki

# Licensed under the MIT license:
# http://www.opensource.org/licenses/mit-license
# Copyright (c) 2011 globo.com thumbor@googlegroups.com

import tempfile
from os.path import expanduser, join

import derpconf.config as config
from derpconf.config import Config

from thumbor import __version__
from thumbor.filters import BUILTIN_FILTERS
from thumbor.handler_lists import BUILTIN_HANDLERS

HOME = expanduser("~")

Config.define("THUMBOR_LOG_CONFIG", None, "Logging configuration as json", "Logging")
Config.define(
    "THUMBOR_LOG_FORMAT",
    "%(asctime)s %(name)s:%(levelname)s %(message)s",
    "Log Format to be used by thumbor when writing log messages.",
    "Logging",
)

Config.define(
    "THUMBOR_LOG_DATE_FORMAT",
    "%Y-%m-%d %H:%M:%S",
    "Date Format to be used by thumbor when writing log messages.",
    "Logging",
)

Config.define(
    "MAX_WIDTH",
    0,
    "Max width in pixels for images read or generated by thumbor",
    "Imaging",
)
Config.define(
    "MAX_HEIGHT",
    0,
    "Max height in pixels for images read or generated by thumbor",
    "Imaging",
)
Config.define(
    "MAX_PIXELS", 75e6, "Max pixel count for images read by thumbor", "Imaging"
)
Config.define(
    "MIN_WIDTH",
    1,
    "Min width in pixels for images read or generated by thumbor",
    "Imaging",
)
Config.define(
    "MIN_HEIGHT",
    1,
    "Min width in pixels for images read or generated by thumbor",
    "Imaging",
)
Config.define(
    "ALLOWED_SOURCES",
    [],
    "Allowed domains for the http loader to download. These are regular expressions.",
    "Imaging",
)
Config.define("QUALITY", 80, "Quality index used for generated JPEG images", "Imaging")
Config.define(
    "PROGRESSIVE_JPEG",
    True,
    "Exports JPEG images with the `progressive` flag set.",
    "Imaging",
)

Config.define(
    "PILLOW_JPEG_SUBSAMPLING",
    None,
    "Specify subsampling behavior for Pillow (see `subsampling` \
              in http://pillow.readthedocs.org/en/latest/handbook/image-file-formats.html#jpeg)."
    'Be careful to use int for 0,1,2 and string for "4:4:4" notation. '
    "Will ignore `quality`. Using `keep` will copy the original file's subsampling.",
    "Imaging",
)

Config.define(
    "PILLOW_JPEG_QTABLES",
    None,
    "Specify quantization tables for Pillow (see `qtables` \
              in http://pillow.readthedocs.org/en/latest/handbook/image-file-formats.html#jpeg). "
    "Will ignore `quality`. Using `keep` will copy the original file's qtables.",
    "Imaging",
)

Config.define(
    "PILLOW_RESAMPLING_FILTER",
    "LANCZOS",
    "Specify resampling filter for Pillow resize method."
    "One of LANCZOS, NEAREST, BILINEAR, BICUBIC, HAMMING, BOX (Pillow>=3.4.0).",
    "Imaging",
)

Config.define(
    "WEBP_QUALITY",
    None,
    "Quality index used for generated WebP images. If not set (None) the same level of "
    "JPEG quality will be used. If 100 the `lossless` flag will be used.",
    "Imaging",
)

Config.define(
    "PNG_COMPRESSION_LEVEL",
    6,
    "Compression level for generated PNG images.",
    "Imaging",
)
Config.define(
    "PILLOW_PRESERVE_INDEXED_MODE",
    True,
    "Indicates if final image should preserve indexed mode (P or 1) of original image",
    "Imaging",
)
Config.define(
    "AUTO_WEBP",
    False,
    "Specifies whether WebP format should be used automatically if the request accepts it "
    "(via Accept header)",
    "Imaging",
)
Config.define(
    "AUTO_PNG_TO_JPG",
    False,
    "Specifies whether a PNG image should be used automatically if the png image has "
    "no transparency (via alpha layer). "
    "WARNING: Depending on case, this is not a good deal. "
    "This transformation maybe causes distortions or the size of image can increase. "
    "Images with texts, for example, the result image maybe will be distorced. "
    "Dark images, for example, the size of result image maybe will be bigger. "
    "You have to evaluate the majority of your use cases "
    "to take a decision about the usage of this conf.",
    "Imaging",
)
Config.define(
    "SVG_DPI",
    150,
    "Specify the ratio between 1in and 1px for SVG images. This is only used when"
    "rasterizing SVG images having their size units in cm or inches.",
    "Imaging",
)
Config.define(
    "MAX_AGE",
    24 * 60 * 60,
    "Max AGE sent as a header for the image served by thumbor in seconds",
    "Imaging",
)
Config.define(
    "MAX_AGE_TEMP_IMAGE",
    0,
    "Indicates the Max AGE header in seconds for "
    "temporary images (images with failed smart detection)",
    "Imaging",
)
Config.define(
    "RESPECT_ORIENTATION",
    False,
    "Indicates whether thumbor should rotate images that have an Orientation EXIF header",
    "Imaging",
)
Config.define(
    "IGNORE_SMART_ERRORS",
    False,
    "Ignore errors during smart detections and return image "
    "as a temp image (not saved in result storage and with MAX_AGE_TEMP_IMAGE age)",
    "Imaging",
)

Config.define(
    "SEND_IF_MODIFIED_LAST_MODIFIED_HEADERS",
    False,
    "Sends If-Modified-Since & Last-Modified headers; requires support from result storage",
    "Imaging",
)

Config.define(
    "PRESERVE_EXIF_INFO",
    False,
    "Preserves exif information in generated images. "
    "Increases image size in kbytes, use with caution.",
    "Imaging",
)

Config.define(
    "PRESERVE_IMAGE_METADATA",
    False,
    "Preserves image metadata (exif, iptc, xmp and comment) in generated images. "
    "Increases image size in kbytes, use with caution.",
    "Imaging",
)

Config.define(
    "ALLOW_ANIMATED_GIFS",
    True,
    "Indicates whether thumbor should enable the EXPERIMENTAL support for animated gifs.",
    "Imaging",
)

Config.define(
    "USE_GIFSICLE_ENGINE",
    False,
    "Indicates whether thumbor should use gifsicle engine. "
    "Please note that smart cropping and filters are not "
    "supported for gifs using gifsicle (but won't give an error).",
    "Imaging",
)

Config.define(
    "USE_BLACKLIST",
    False,
    "Indicates whether thumbor should enable blacklist "
    "functionality to prevent processing certain images.",
    "Imaging",
)

Config.define(
    "ENGINE_THREADPOOL_SIZE",
    0,
    "Size of the thread pool used for image transformations."
    " The default value is 0 (don't use a threadpoool. "
    "Increase this if you are seeing your IOLoop getting "
    "blocked (often indicated by your upstream HTTP "
    "requests timing out)",
    "Imaging",
)

Config.define(
    "METRICS",
    "thumbor.metrics.logger_metrics",
    "The metrics backend thumbor should use to measure internal actions."
    " This must be the full name of a python module "
    "(python must be able to import it)",
    "Extensibility",
)
Config.define(
    "LOADER",
    "thumbor.loaders.http_loader",
    "The loader thumbor should use to load the original image. "
    "This must be the full name of a python module "
    "(python must be able to import it)",
    "Extensibility",
)
Config.define(
    "STORAGE",
    "thumbor.storages.file_storage",
    "The file storage thumbor should use to store original images."
    " This must be the full name of a python module "
    "(python must be able to import it)",
    "Extensibility",
)
Config.define(
    "RESULT_STORAGE",
    None,
    "The result storage thumbor should use to store generated "
    "images. This must be the full name of a python "
    "module (python must be able to import it)",
    "Extensibility",
)
Config.define(
    "ENGINE",
    "thumbor.engines.pil",
    "The imaging engine thumbor should use to perform image "
    "operations. This must be the full name of a "
    "python module (python must be able to import it)",
    "Extensibility",
)

Config.define(
    "GIF_ENGINE",
    "thumbor.engines.gif",
    "The gif engine thumbor should use to perform image operations."
    " This must be the full name of a "
    "python module (python must be able to import it)",
    "Extensibility",
)

Config.define(
    "SECURITY_KEY",
    "MY_SECURE_KEY",
    "The security key thumbor uses to sign image URLs",
    "Security",
)

Config.define(
    "ALLOW_UNSAFE_URL",
    True,
    "Indicates if the /unsafe URL should be available",
    "Security",
)
Config.define("ENABLE_ETAGS", True, "Enables automatically generated etags", "HTTP")
Config.define(
    "MAX_ID_LENGTH", 32, "Set maximum id length for images when stored", "Storage",
)
Config.define(
    "GC_INTERVAL", None, "Set garbage collection interval in seconds", "Performance",
)

Config.define(
    "HEALTHCHECK_ROUTE", r"/healthcheck/?", "Healthcheck route.", "Healthcheck"
)

# METRICS OPTIONS
Config.define("STATSD_HOST", None, "Host to send statsd instrumentation to", "Metrics")
Config.define("STATSD_PORT", 8125, "Port to send statsd instrumentation to", "Metrics")
Config.define("STATSD_PREFIX", None, "Prefix for statsd", "Metrics")

# FILE LOADER OPTIONS
Config.define(
    "FILE_LOADER_ROOT_PATH",
    HOME,
    "The root path where the File Loader will try to find images",
    "File Loader",
)

# HTTP LOADER OPTIONS
Config.define(
    "HTTP_LOADER_CONNECT_TIMEOUT",
    5,
    "The maximum number of seconds libcurl can take to connect to an image being loaded",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_REQUEST_TIMEOUT",
    20,
    "The maximum number of seconds libcurl can take to download an image",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_FOLLOW_REDIRECTS",
    True,
    "Indicates whether libcurl should follow redirects when downloading an image",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_MAX_REDIRECTS",
    5,
    "Indicates the number of redirects libcurl should follow when downloading an image",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_MAX_CLIENTS",
    10,
    "The maximum number of simultaneous HTTP connections the loader can make before queuing",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_FORWARD_USER_AGENT",
    False,
    "Indicates whether thumbor should forward the user agent of the requesting user",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_FORWARD_ALL_HEADERS",
    False,
    "Indicates whether thumbor should forward the headers of the request",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_FORWARD_HEADERS_WHITELIST",
    [],
    "Indicates which headers should be forwarded among all the headers of the request",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_DEFAULT_USER_AGENT",
    "Thumbor/%s" % __version__,
    "Default user agent for thumbor http loader requests",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_PROXY_HOST",
    None,
    "The proxy host needed to load images through",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_PROXY_PORT", None, "The proxy port for the proxy host", "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_PROXY_USERNAME",
    None,
    "The proxy username for the proxy host",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_PROXY_PASSWORD",
    None,
    "The proxy password for the proxy host",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_CA_CERTS",
    None,
    "The filename of CA certificates in PEM format",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_VALIDATE_CERTS",
    None,
    "Validate the server’s certificate for HTTPS requests",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_CLIENT_KEY", None, "The filename for client SSL key", "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_CLIENT_CERT",
    None,
    "The filename for client SSL certificate",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_CURL_ASYNC_HTTP_CLIENT",
    False,
    "If the CurlAsyncHTTPClient should be used",
    "HTTP Loader",
)
Config.define(
    "HTTP_LOADER_CURL_LOW_SPEED_TIME",
    0,
    "If HTTP_LOADER_CURL_LOW_SPEED_LIMIT and HTTP_LOADER_CURL_ASYNC_HTTP_CLIENT "
    + "are set, then this is the time in seconds as integer after a download should "
    + "timeout if the speed is below HTTP_LOADER_CURL_LOW_SPEED_LIMIT for that long",
)
Config.define(
    "HTTP_LOADER_CURL_LOW_SPEED_LIMIT",
    0,
    "If HTTP_LOADER_CURL_LOW_SPEED_TIME and HTTP_LOADER_CURL_ASYNC_HTTP_CLIENT "
    + "are set, then this is the limit in bytes per second as integer which should "
    + "timeout if the speed is below that limit for HTTP_LOADER_CURL_LOW_SPEED_TIME seconds",
)

# FILE STORAGE GENERIC OPTIONS
Config.define(
    "STORAGE_EXPIRATION_SECONDS",
    60 * 60 * 24 * 30,
    "Expiration in seconds for the images in the File Storage. Defaults to one month",
    "File Storage",
)
Config.define(
    "STORES_CRYPTO_KEY_FOR_EACH_IMAGE",
    False,
    "Indicates whether thumbor should store the signing key for each image in the file storage. "
    + "This allows the key to be changed and old images to still be properly found",
    "File Storage",
)

# FILE STORAGE OPTIONS
Config.define(
    "FILE_STORAGE_ROOT_PATH",
    join(tempfile.gettempdir(), "thumbor", "storage"),
    "The root path where the File Storage will try to find images",
    "File Storage",
)

# PHOTO UPLOAD OPTIONS
Config.define(
    "UPLOAD_MAX_SIZE", 0, "Max size in bytes for images uploaded to thumbor", "Upload",
)
Config.define(
    "UPLOAD_ENABLED",
    False,
    "Indicates whether thumbor should enable File uploads",
    "Upload",
)
Config.define(
    "UPLOAD_PHOTO_STORAGE",
    "thumbor.storages.file_storage",
    "The type of storage to store uploaded images with",
    "Upload",
)
Config.define(
    "UPLOAD_DELETE_ALLOWED",
    False,
    "Indicates whether image deletion should be allowed",
    "Upload",
)
Config.define(
    "UPLOAD_PUT_ALLOWED",
    False,
    "Indicates whether image overwrite should be allowed",
    "Upload",
)
Config.define(
    "UPLOAD_DEFAULT_FILENAME", "image", "Default filename for image uploaded", "Upload",
)

# ALIASES FOR OLD PHOTO UPLOAD OPTIONS
Config.alias("MAX_SIZE", "UPLOAD_MAX_SIZE")
Config.alias("ENABLE_ORIGINAL_PHOTO_UPLOAD", "UPLOAD_ENABLED")
Config.alias("ORIGINAL_PHOTO_STORAGE", "UPLOAD_PHOTO_STORAGE")
Config.alias("ALLOW_ORIGINAL_PHOTO_DELETION", "UPLOAD_DELETE_ALLOWED")
Config.alias("ALLOW_ORIGINAL_PHOTO_PUTTING", "UPLOAD_PUT_ALLOWED")

# MIXED STORAGE OPTIONS
Config.define(
    "MIXED_STORAGE_FILE_STORAGE",
    "thumbor.storages.no_storage",
    "Mixed Storage file storage. This must be the full name "
    "of a python module (python must be able to import it)",
    "Mixed Storage",
)
Config.define(
    "MIXED_STORAGE_CRYPTO_STORAGE",
    "thumbor.storages.no_storage",
    "Mixed Storage signing key storage. This must be the full "
    "name of a python module (python must be able to import it)",
    "Mixed Storage",
)
Config.define(
    "MIXED_STORAGE_DETECTOR_STORAGE",
    "thumbor.storages.no_storage",
    "Mixed Storage detector information storage. This must be the full "
    "name of a python module (python must be able to import it)",
    "Mixed Storage",
)

# JSON META ENGINE OPTIONS
Config.define(
    "META_CALLBACK_NAME",
    None,
    "The callback function name that should be used by the META route for JSONP access",
    "Meta",
)

# DETECTORS OPTIONS
Config.define(
    "DETECTORS",
    [],
    "List of detectors that thumbor should use to find faces and/or features. All of them must be "
    + "full names of python modules (python must be able to import it)",
    "Detection",
)

# FACE DETECTOR CASCADE FILE
Config.define(
    "FACE_DETECTOR_CASCADE_FILE",
    "haarcascade_frontalface_alt.xml",
    "The cascade file that opencv will use to detect faces.",
    "Detection",
)

Config.define(
    "GLASSES_DETECTOR_CASCADE_FILE",
    "haarcascade_eye_tree_eyeglasses.xml",
    "The cascade file that opencv will use to detect glasses.",
    "Detection",
)

Config.define(
    "PROFILE_DETECTOR_CASCADE_FILE",
    "haarcascade_profileface.xml",
    "The cascade file that opencv will use to detect profile faces.",
    "Detection",
)

Config.define(
    "OPTIMIZERS",
    [
        # 'thumbor.optimizers.jpegtran',
        # 'thumbor.optimizers.gifv',
    ],
    "List of optimizers that thumbor will use to optimize images",
    "Optimizers",
)

# OPTIMIZER CONFIGURATIONS
Config.define(
    "JPEGTRAN_PATH", "/usr/bin/jpegtran", "Path for the jpegtran binary", "Optimizers",
)

Config.define(
    "JPEGTRAN_SCANS_FILE",
    "",
    "Path for the progressive scans file to use with "
    "jpegtran optimizer. Implies progressive jpeg output",
    "Optimizers",
)

Config.define(
    "FFMPEG_PATH",
    "/usr/local/bin/ffmpeg",
    "Path for the ffmpeg binary used to generate gifv(h.264)",
    "Optimizers",
)

# AVAILABLE FILTERS
Config.define(
    "FILTERS",
    BUILTIN_FILTERS,
    "List of filters that thumbor will allow to be used in generated images. All of them must be "
    + "full names of python modules (python must be able to import it)",
    "Filters",
)

# RESULT STORAGE
Config.define(
    "RESULT_STORAGE_EXPIRATION_SECONDS",
    0,
    "Expiration in seconds of generated images in the result storage",
    "Result Storage",
)  # Never expires
Config.define(
    "RESULT_STORAGE_FILE_STORAGE_ROOT_PATH",
    join(tempfile.gettempdir(), "thumbor", "result_storage"),
    "Path where the Result storage will store generated images",
    "Result Storage",
)
Config.define(
    "RESULT_STORAGE_STORES_UNSAFE",
    False,
    "Indicates whether unsafe requests should also be stored in the Result Storage",
    "Result Storage",
)

# QUEUED DETECTOR REDIS OPTIONS
Config.define(
    "REDIS_QUEUE_SERVER_HOST",
    "localhost",
    "Server host for the queued redis detector",
    "Queued Redis Detector",
)
Config.define(
    "REDIS_QUEUE_SERVER_PORT",
    6379,
    "Server port for the queued redis detector",
    "Queued Redis Detector",
)
Config.define(
    "REDIS_QUEUE_SERVER_DB",
    0,
    "Server database index for the queued redis detector",
    "Queued Redis Detector",
)
Config.define(
    "REDIS_QUEUE_SERVER_PASSWORD",
    None,
    "Server password for the queued redis detector",
    "Queued Redis Detector",
)

# QUEUED DETECTOR SQS OPTIONS
Config.define("SQS_QUEUE_KEY_ID", None, "AWS key id", "Queued SQS Detector")
Config.define("SQS_QUEUE_KEY_SECRET", None, "AWS key secret", "Queued SQS Detector")
Config.define("SQS_QUEUE_REGION", "us-east-1", "AWS SQS region", "Queued SQS Detector")

# ERROR HANDLING
Config.define(
    "USE_CUSTOM_ERROR_HANDLING",
    False,
    "This configuration indicates whether thumbor should use a custom error handler.",
    "Errors",
)
Config.define(
    "ERROR_HANDLER_MODULE",
    "thumbor.error_handlers.sentry",
    "Error reporting module. Needs to contain a class called ErrorHandler with a "
    + "handle_error(context, handler, exception) method.",
    "Errors",
)

# SENTRY REPORTING MODULE
Config.define(
    "SENTRY_DSN_URL",
    "",
    "Sentry thumbor project dsn. i.e.: "
    + "http://5a63d58ae7b94f1dab3dee740b301d6a:73eea45d3e8649239a973087e8f21f98@localhost:9000/2",
    "Errors - Sentry",
)

# SENTRY REPORTING MODULE
Config.define(
    "SENTRY_ENVIRONMENT", None, "Sentry environment i.e.: staging ", "Errors - Sentry",
)

# FILE REPORTING MODULE
Config.define("ERROR_FILE_LOGGER", None, "File of error log as json", "Errors")
Config.define(
    "ERROR_FILE_NAME_USE_CONTEXT",
    False,
    "File of error log name is parametrized with context attribute",
    "Errors",
)

# SIGNER MODULE
Config.define(
    "URL_SIGNER",
    "libthumbor.url_signers.base64_hmac_sha1",
    "The url signer thumbor should use to verify url signatures."
    + "This must be the full name of a python module "
    + "(python must be able to import it)",
    "Extensibility",
)

# SERVER
Config.define(
    "MAX_WAIT_SECONDS_BEFORE_SERVER_SHUTDOWN",
    0,
    "The amount of time to wait before shutting down the server, i.e. stop accepting requests.",
    "Server",
)
Config.define(
    "MAX_WAIT_SECONDS_BEFORE_IO_SHUTDOWN",
    0,
    "The amount of time to waut before shutting down all io, after the server has been stopped",
    "Server",
)

# HANDLER LISTS
Config.define(
    "HANDLER_LISTS",
    BUILTIN_HANDLERS,
    "Handler Lists are responsible for adding new handlers to thumbor app.",
    "HandlerLists",
)


Config.define(
    "APP_CLASS",
    "thumbor.app.ThumborServiceApp",
    "Custom app class to override ThumborServiceApp. "
    "This config value is overridden by the -a command-line parameter.",
)


def generate_config():
    config.generate_config()


def format_value(value):
    if isinstance(value, str):
        return "'%s'" % value
    if isinstance(value, (tuple, list, set)):
        representation = "[\n"
        for item in value:
            representation += "#    %s" % item
        representation += "#]"
        return representation
    return value


if __name__ == "__main__":
    generate_config()
